### PR TITLE
Automatically assign a midi input device to the selected track

### PR DESCRIPTION
**Pull request by LNooteboom**
Hello everyone, this is my first pullrequest here!

This pullrequest implements "Faster access to MIDI assignment/auto-assignment" in meta-issue #1472 (as well as #3400 #1189 #1781 #1918 #1942 #3185)

With this pullrequest, it is now possible to automatically assign/connect a "master" midi input device (eg. a keyboard) to the currently selected instrument track and piano roll, this makes it behave similarly to the QWERTY keyboard for note input.
Whereas previously you needed to manually assign your MIDI keyboard via the menu every time you wanted to play a different instrument, now you can just select your main midi keyboard in the settings menu (MIDI Tab), and LMMS will automatically assign it. This speeds up the workflow for people who record using a MIDI device.

The use of this feature is completely optional, if you leave the auto-assign setting at 'none', your workflow will not be impacted at all.

### CODE REVIEW

**Comment by PhysSong** [comment]
Thanks for the great work!
Some thoughts:
- The MIDI input gets unassigned if I open a dialog or a context menu. It'd be great if we could improve that behavior.
- This is useful with sequenced MIDI clients, but not with raw MIDI clients, i.e., MIDI backends which don't separate inputs from different MIDI input devices.

File: src/gui/SetupDialog.cpp [hunk 1]
```diff
@@ -685,7 +685,14 @@ SetupDialog::SetupDialog(ConfigTabs tab_to_open) :
...
 	m_assignableMidiDevices->setGeometry(10, 20, 240, 28);
 	m_assignableMidiDevices->addItem("none");
-	m_assignableMidiDevices->addItems(Engine::mixer()->midiClient()->readablePorts());
+	if ( !Engine::mixer()->midiClient()->isRaw() )
+	{
+		m_assignableMidiDevices->addItems(Engine::mixer()->midiClient()->readablePorts());
+	}
+	else
+	{
+		m_assignableMidiDevices->addItem("yes");
```
Comment (PhysSong) [inline src/gui/SetupDialog.cpp:694]: "none" vs. "yes" look a bit odd. I think "all" might be better, any ideas?

**Comment by qnebra** [comment]
Works fine for casual preview and recording with midi devices. Of course, automation was bit tricker and still requiered more steps, but overall this is fine.

**Review by PhysSong — APPROVED** [review]

**Comment by qnebra** [comment]
You are in 1.2.2 and this feature is in at least 1.3-alpha

**Comment by Saitama697** [comment]
> You are in 1.2.2 and this feature is in at least 1.3-alpha

Okay, thank you! I’m just curious— the post I quoted was made on May 15, 2020. How long does it usually take for an update to reach the stable version?

Just curious.

Thank you!

**Pull request merged**